File: c-sharp/RedisShim/CountMinSketchShim.cs
using System;
using System.Threading.Tasks;
using StackExchange.Redis;

namespace count_min_sketch.RedisShim
{
    public class CountMinSketchShim
    {
        private readonly ConnectionMultiplexer redisConn;

        // StackOverflow.Redis does not have native RedisBloom support.
        // So, I created this shim using it's support for direct execution of arbitrary commands.
        public CountMinSketchShim(ConnectionMultiplexer redisConn)
        {
            this.redisConn = redisConn;
        }

        public string InitByDim(string key, int width, int depth)
        {
            return (string)redisConn.GetDatabase().Execute("CMS.INITBYDIM", key, width, depth);
        }

        public string InitByProb(string key, float error, float probability)
        {
            return (string)redisConn.GetDatabase().Execute("CMS.INITBYPROB", key, error, probability);
        }

        public RedisResult IncrBy(string key, string item, int increment=1)
        {
            return redisConn.GetDatabase().Execute("CMS.INCRBY", key, item, increment);;
        }

        public RedisResult Query(string key, string item)
        {
            return redisConn.GetDatabase().Execute("CMS.QUERY", key, item);;
        }

        public RedisResult Merge(string dest, string[] sources, int[] weights)
        {
            if (sources.Length != weights.Length)
            {
                throw new ArgumentException("sources.Length must equal weights.Length");
            }
            
            return redisConn.GetDatabase().Execute("CMS.MERGE", dest, sources.Length, sources, "WEIGHTS", weights);
        }

        public Task<RedisResult> MergeAsync(string dest, string[] sources, int[] weights)
        {
            if (sources.Length != weights.Length)
            {
                throw new ArgumentException("sources.Length must equal weights.Length");
            }
            
            return redisConn.GetDatabase().ExecuteAsync("CMS.MERGE", dest, sources.Length, sources, "WEIGHTS", weights);
        }
    }
}

File: src/info/jackpines/redisshim/CountMinSketchShim.java
package info.jackpines.redisshim;

import java.nio.charset.StandardCharsets;
import redis.Command;
import redis.client.RedisClientBase;
import redis.reply.MultiBulkReply;
import redis.reply.StatusReply;

public class CountMinSketchShim {
  private final RedisClientBase client;

  public CountMinSketchShim( RedisClientBase client ) {
    this.client = client;
  }

  private static final String INITBYDIM = "CMS.INITBYDIM";
  private static final byte[] INITBYDIM_BYTES = INITBYDIM.getBytes( StandardCharsets.US_ASCII );

  public String initByDim( String key, int width, int depth ) {
    return ( (StatusReply) client.execute( INITBYDIM, new Command( INITBYDIM_BYTES, key, width, depth ) ) ).data();
  }

  private static final String INITBYPROB = "CMS.INITBYPROB";
  private static final byte[] INITBYPROB_BYTES = INITBYPROB.getBytes( StandardCharsets.US_ASCII );

  public String initByProb( String key, float error, float probability ) {
    return ( (StatusReply) client.execute( INITBYPROB, new Command( INITBYPROB_BYTES, key, error, probability ) ) )
        .data();
  }

  private static final String INCRBY = "CMS.INCRBY";
  private static final byte[] INCRBY_BYTES = INCRBY.getBytes(StandardCharsets.US_ASCII);

  public long incrBy( String key, String item ) {
    return incrBy( key, item, 1 );
  }

  public long incrBy( String key, String item, int increment ) {
    return Long.parseLong(
        ( (MultiBulkReply) client.execute( INCRBY, new Command( INCRBY_BYTES, key, item, increment ) ) )
            .asStringList( StandardCharsets.UTF_8 ).get( 0 ) );
  }

  private static final String QUERY = "CMS.QUERY";
  private static final byte[] QUERY_BYTES = QUERY.getBytes( StandardCharsets.US_ASCII );

  public long query( String key, String item ) {
    return Long.parseLong( ( (MultiBulkReply) client.execute( QUERY, new Command( QUERY_BYTES, key, item ) ) )
        .asStringList( StandardCharsets.UTF_8 ).get( 0 ) );
  }
}
